xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_fog.cpp

```diff
@@ -1,8 +1,309 @@
 #include "stdafx.h"
 #include "rasterizer_dx9_fog.h"
 
+#include "rasterizer_dx9.h"
+#include "rasterizer_dx9_fullscreen_passes.h"
+#include "rasterizer_dx9_main.h"
+#include "rasterizer_dx9_submit.h"
+#include "rasterizer_dx9_shader_submit.h"
+
+#include "rasterizer/rasterizer_globals.h"
+#include "rasterizer/dx9/shaders/compiled/fog_atmospheric_apply.h"
+
+/* constants */
+
 static real32 fog_wind_x_axis_velocity_factor1 = 30.0f;
 
+/* typedefs */
+
+typedef bool(__cdecl* rasterizer_dx9_create_fog_shaders_t)(void);
+
+/* globals */
+
+rasterizer_dx9_create_fog_shaders_t p_rasterizer_dx9_create_fog_shaders;
+
+/* prototypes */
+
+real32* patchy_fog_apply_scale_get(void);
+bool __cdecl rasterizer_dx9_set_fog_pixel_shader(uint32 index);
+void rasterizer_fog_constant_velocity_update();
+void rasterizer_fog_speed_apply_patch(void);
+bool __cdecl rasterizer_dx9_create_fog_shaders(void);
+void rasterizer_dx9_atmospheric_fog_patch(void);
+
+/* public code */
+
+void rasterizer_dx9_fog_apply_patches(void)
+{
+	rasterizer_fog_speed_apply_patch();
+
+	// splitscreen patchy fog fix, take just the viewport portion of the main render target
+	// that is used as a texture in the patchy fog draw code shader
+	WritePointer(Memory::GetAddress(0x2774AE) + 1, rasterizer_dx9_patchy_fog_apply_build_vertex_buffer);
+	WritePointer(Memory::GetAddress(0x2774CF) + 1, rasterizer_dx9_patchy_fog_apply_from_stencil_build_vertex_buffer);
+
+
+	rasterizer_dx9_atmospheric_fog_patch();
+	return;
+}
+
+bool __cdecl rasterizer_dx9_draw_atmospheric_fog(int32 a1)
```

Comment:
this sets up the pipeline, doesn't render anything at all

Reply:
function name changed to `rasterizer_dx9_atmospheric_fog_pipeline_setup`